### DISCUSSION TITLE
How to update container without downtime?

### DISCUSSION THREAD

**Discussion by qupfer** [containers/podman#13348] [Q&A]
Hi, is there a "default" way to update container without (or very short downtime)?

Of cause, the easy way to upgrade is pull;stop;rm;up.
For docker-compose. there seems a ways with repicas.
1 pull
2 set repicas (--scale) to 2
3 start with --no-recreate 
4 remove old
5 set scale to 1 again

Is this or something similar possible with podman?

**Comment by vrothberg** [comment]
Thanks for reaching out, @qupfer!

Regarding "no downtime": no, that is not possible with Podman.  Podman does not support replicas since that would require more than one machine/node.  Podman's sole focus is on single-node use cases.  Everything that goes beyond that is Kubernetes' territory.

Regarding "short downtime": I recommend having a look at [podman-auto-update](https://www.redhat.com/sysadmin/podman-auto-updates-rollbacks).

**Comment by eriksjolund** [comment]
@qupfer "_without downtime_" could mean different things.
One interpretation could be: A client would not notice a restart of a containerized server. (The container image could be updated during the restart).

Disclaimer: I'm still learning how this all works, but this my current understanding of the topic.

If these statements are all true

* There is a containerized server application that was started by Podman via _socket activation_. (Let's call the containerized server application _myserver_).
* _myserver_  accepted a new connection (by running [accept()](https://man7.org/linux/man-pages/man2/accept.2.html)).
* _myserver_ stored the accepted socket with [FDSTORE](https://www.freedesktop.org/software/systemd/man/sd_notify.html#FDSTORE=1).
* _myserver_ stored the application state with `FDSTORE`.  _Application state can either be serialized to a file in /run/, or better, stored in a  memory file descriptor. memfd_create(2)_ quote from [man sd_notify](https://www.freedesktop.org/software/systemd/man/sd_notify.html#FDSTORE=1) (See also [memfd_create(2)](http://man7.org/linux/man-pages/man2/memfd_create.2.html)).
* The _myserver_ crashed or was explicitly restarted (e.g. `systemctl --user restart myserver.service`). _This is useful for implementing services that can restart after an explicit request or a crash without losing state_ quote from [man sd_notify](https://www.freedesktop.org/software/systemd/man/sd_notify.html#FDSTORE=1).

then the client would not notice a restart of _myserver_, because the listening socket FD, the accepted socket FD and the application state FD could be passed via `LISTEN_FDS` and `LISTEN_FDNAMES` (see [man sd_listen_fds](https://www.freedesktop.org/software/systemd/man/sd_listen_fds.html)) to a newly started _myserver_. The application state could therefore be restored.

Please correct me if I've misunderstood anything.

It would be cool to see if this could be made to work!

A side-note: I wrote a demo [mariadb-podman-socket-activation](https://github.com/eriksjolund/mariadb-podman-socket-activation/) where Podman starts a MariaDB container with _socket activation_. The demo does not make use of `FDSTORE`, though.
